src/mongoose_acc.erl

```diff
@@ -1,300 +1,263 @@
-%% persistent stuff you want to pass between processes there.
--spec strip(t()) -> t().
-strip(Acc) ->
-    strip(Acc, get(element, Acc)).
-
-%% @doc alt version in case we want to replace the xmlel we are sending
--spec strip(t(), exml:element()) -> t().
-strip(Acc, El) ->
-    Acc1 = from_element(El),
-    Attributes = [ref, timestamp],
-    NewAcc = lists:foldl(fun(Attrib, AccIn) ->
-                           Val = maps:get(Attrib, Acc),
-                           maps:put(Attrib, Val, AccIn)
+-type new_acc_params() :: #{
+        location := location(),
+        lserver := jid:lserver(),
+        element := exml:element() | undefined,
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid() % optional
+       }.
+
+-type strip_params() :: #{
+        lserver := jid:lserver(),
+        element := exml:element(),
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid() % optional
+       }.
+
+-type stanza_params() :: #{
+        element := exml:element() | undefined,
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid(), % optional
+        _ => _
+       }.
+
+-type ns_key() :: {NS :: any(), Key :: any()}.
+
+-type strippable() :: boolean() | undefined.
+
+%% --------------------------------------------------------
+%% API
+%% --------------------------------------------------------
+
+-spec new(Params :: new_acc_params()) -> t().
+new(#{ location := Location, lserver := LServer } = Params) ->
+    Stanza = stanza_from_params(Params),
+    #{
+      mongoose_acc => true,
+      ref => make_ref(),
+      timestamp => os:timestamp(),
+      origin_pid => self(),
+      origin_location => Location,
+      origin_stanza => Stanza, % Stanza that triggered the processing
+      stanza => Stanza,
+      lserver => LServer,
+      non_strippable => sets:new()
+     }.
+
+ref(#{ mongoose_acc := true, ref := Ref }) ->
+    Ref.
+
+timestamp(#{ mongoose_acc := true, timestamp := TS }) ->
+    TS.
+
+lserver(#{ mongoose_acc := true, lserver := LServer }) ->
+    LServer.
+
+element(#{ mongoose_acc := true, stanza := #{ element := El } }) ->
+    El;
+element(#{ mongoose_acc := true }) ->
+    undefined.
+
+from_jid(#{ mongoose_acc := true, stanza := #{ from_jid := FromJID } }) ->
+    FromJID;
+from_jid(#{ mongoose_acc := true }) ->
+    undefined.
+
+to_jid(#{ mongoose_acc := true, stanza := #{ to_jid := ToJID } }) ->
+    ToJID;
+to_jid(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_name(#{ mongoose_acc := true, stanza := #{ name := Name } }) ->
+    Name;
+stanza_name(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_type(#{ mongoose_acc := true, stanza := #{ type := Type } }) ->
+    Type;
+stanza_type(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_ref(#{ mongoose_acc := true, stanza := #{ ref := StanzaRef } }) ->
+    StanzaRef;
+stanza_ref(#{ mongoose_acc := true }) ->
+    undefined.
+
+%% While possible, it is not recommended to update stanza to 'undefined'.
```

Comment:
Would it be worth to display a warning when the stanza is updated to `undefined`?

Reply:
Not really, it should be actually resolved in iteration 2 to get rid of accumulator in places where there is no "stanza".